java-checks/src/main/java/org/sonar/java/checks/RedundantThrowsDeclarationCheck.java

```diff
@@ -37,32 +56,166 @@
-      String fullyQualifiedName = symbolType.fullyQualifiedName();
+      Type exceptionType = typeTree.symbolType();
+      String fullyQualifiedName = exceptionType.fullyQualifiedName();
       if (!reported.contains(fullyQualifiedName)) {
-        String superTypeName = isSubclassOfAny(symbolType, thrownList);
+        String superTypeName = isSubclassOfAny(exceptionType, thrownList);
         if (superTypeName != null) {
-          reportIssue(typeTree, "Remove the declaration of thrown exception '" + fullyQualifiedName + "' which is a subclass of '" + superTypeName + "'.");
-        } else if (symbolType.isSubtypeOf("java.lang.RuntimeException")) {
-          reportIssue(typeTree, "Remove the declaration of thrown exception '" + fullyQualifiedName + "' which is a runtime exception.");
+          reportIssue(typeTree, String.format("Remove the declaration of thrown exception '%s' which is a subclass of '%s'.", fullyQualifiedName, superTypeName));
+        } else if (exceptionType.isSubtypeOf("java.lang.RuntimeException")) {
+          reportIssue(typeTree, String.format("Remove the declaration of thrown exception '%s' which is a runtime exception.", fullyQualifiedName));
         } else if (declaredMoreThanOnce(fullyQualifiedName, thrownList)) {
-          reportIssue(typeTree, "Remove the redundant '" + fullyQualifiedName + "' thrown exception declaration(s).");
+          reportIssue(typeTree, String.format("Remove the redundant '%s' thrown exception declaration(s).", fullyQualifiedName));
+        } else if (canNotBeThrown(methodTree, exceptionType, thrownExceptions)) {
+          reportIssue(typeTree, String.format("Remove the declaration of thrown exception '%s', as it cannot be thrown from %s's body.", fullyQualifiedName,
+            tree.is(Tree.Kind.CONSTRUCTOR) ? "constructor" : "method"));
         }
         reported.add(fullyQualifiedName);
       }
     }
   }
 
+  private static boolean canNotBeThrown(MethodTree methodTree, Type exceptionType, @Nullable Set<Type> thrownExceptions) {
+    if (isOverridingOrDesignedForExtension(methodTree)
+      || !exceptionType.isSubtypeOf("java.lang.Exception")
+      || exceptionType.isSubtypeOf("java.lang.RuntimeException")
+      || thrownExceptions == null) {
+      return false;
+    }
+
+    if (thrownExceptions.stream().anyMatch(t -> ((JavaType) t).isTagged(JavaType.TYPEVAR))) {
+      // kill the noise due to SONARJAVA-1778 - type substitution not applied on thrown type when parameterized on parametric methods
+      return false;
+    }
+
+    return thrownExceptions.stream().noneMatch(t -> t.isSubtypeOf(exceptionType));
+  }
+
+  private static boolean isOverridingOrDesignedForExtension(MethodTree methodTree) {
+    // we need to be sure that it's not an override
+    return !Boolean.FALSE.equals(((MethodTreeImpl) methodTree).isOverriding())
+      || SERIALIZABLE_CONTRACT_METHODS.contains(methodTree.simpleName().name())
+      // kill the noise - usually designed for extension
```

Comment:
Unecessary comment IMO : I don't get it right away anyway. And I think if it deserves a comment it should be on the method itself, not the call. 
Moreover : kill the noise is a term used internally. should not be part of code IMO.